src/scores/probability/fss_impl.py

```diff
@@ -0,0 +1,528 @@
+f8x3 = np.dtype("f8, f8, f8")
+# Note: `TypeAlias` on variables for python <=3.9
+if TYPE_CHECKING:  # pragma: no cover
+    FssDecomposed = Union[np.ArrayLike, np.DtypeLike]
+else:  # pragma: no cover
+    FssDecomposed = npt.NDArray[f8x3]
+
+
+class DimensionError(ValueError):
+    """
+    Wrapper for specific coordinate/dimension mismatch errors
+    """
+
+    # better to be explicit because python doesn't have braces for scoping its
+    # control flow
+    pass  # pylint: disable=unnecessary-pass
+
+
+class FssComputeMethod(Enum):
+    """
+    Choice of compute backend for FSS.
+    - Standard
+        - NUMPY (default)
+    - Optimized
+        - NUMBA_NATIVE (unavailable)
+        - NUMBA_PARALLEL (unavailable)
+    - Experimental
+        - RUST_NATIVE (unavailable)
+        - RUST_OCL (unavailable)
+    """
+
+    NUMPY = 1  # (default)
+    NUMBA_NATIVE = 2
+    NUMBA_PARALLEL = 3
+    RUST_NATIVE = 4
+    RUST_OCL = 5
+
+
+def fss_2d(
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    *,  # Force keywords arguments to be keyword-only
+    threshold: np.float64,
+    window: Tuple[int, int],
+    spatial_dims: Tuple[str, str],
+    zero_padding: bool = False,
+    reduce_dims: Optional[Iterable[str]] = None,
+    preserve_dims: Optional[Iterable[str]] = None,
+    compute_method: FssComputeMethod = FssComputeMethod.NUMPY,
+    dask: str = "forbidden",  # see: `xarray.apply_ufunc` for options
+) -> XarrayLike:  # pylint: disable=too-many-locals
+    """
+    Aggregates the Fractions Skill Score (FSS) over 2-D spatial coordinates specified by
+    `spatial_dims`.
+
+    Optionally aggregates the output along other dimensions if `reduce_dims` or
+    (mutually exclusive) `preserve_dims` are specified.
+
+    For implementation for a single 2-D field see: :py:func:`fss_2d_single_field`
+
+    Args:
+        fcst: An array of forecasts
+        obs: An array of observations (same spatial shape as `fcst`)
+        threshold: A scalar to compare `fcst` and `obs` fields to generate a
+            binary "event" field.
+        window: A pair of positive integers `(height, width)` of the sliding
+            window; the window dimensions must be greater than 0 and fit within
+            the shape of `obs` and `fcst`.
+        spatial_dims: A pair of dimension names `(x, y)` where `x` and `y` are
+            the spatial dimensions to  slide the window along to compute the FSS.
+            e.g. `("lat", "lon")`.
+        zero_padding: If set to true, applies a 0-valued window border around the
+            data field before computing the FSS. If set to false (default), it
+            uses the edges (thickness = window size) of the input fields as the border.
+            One can think of it as:
+            - zero_padding = False => inner window
+            - zero_padding = True => outer window with 0 values
+        reduce_dims: Optional set of additional dimensions to aggregate over
+            (mutually exclusive to `reduce_dims`).
```

Comment:
...mutually exclusive to `preserve_dims`...

Reply:
😆 good catch